File: julia/HyperTranscend.jl
# ⚠️ AI-assisted; verify. / 生成AI使用・要検証
"""
HyperTranscend — an **experimental** unified computation of transcendental functions
(exp, log, sqrt, ^) for a hypercomplex number of any dimension M = 2^k, via the LEFT
regular representation.  These are *function values defined through L_x*, not a claim
that every hypercomplex analytic function is captured; which algebraic identities hold
is CHECKED per dimension in `self_test()`, not assumed (for non-associative M ≥ 8 the
branch of a matrix function, and left- vs right-functions, need case-by-case care).

  Core identity (this session's result):
      a hypercomplex element x  ≙  its left-multiplication matrix  L_x  (M×M),
      and for any analytic f,   f(x) = f(L_x) · e0.
  So *, ^, exp, log, sqrt are ONE recipe with M as the only knob; the scalar (M=1)
  is the 1×1 case.

  Forward ops (*, exp) are total for every input, zero divisors included — the tensor
  rule never has an exception.  Only INVERSE-type ops (/, log, sqrt, x^negative/fractional)
  can break, and they break in exactly one place: when L_x is singular (a zero divisor,
  which first appears at M ≥ 16).  There we do NOT emit NaN — we name it with a flag.

  Flags (per number): SING = L_x singular ⇒ no unique inverse (zero divisor);
                      CPLX = result left the reals (imag residue) ⇒ go to a bigger field;
                      OVER = a component saturated to ±MAX (range overflow).
"""
module HyperTranscend

using LinearAlgebra
import Random
export Hyper, hexp, hlog, hsqrt, e0, isreal_ok, flags, SING, CPLX, OVER

const SING = 0x01
const CPLX = 0x02
const OVER = 0x04
const MAXF = floatmax(Float64)

struct Hyper
    c::Vector{Float64}
    flag::UInt8
end
Hyper(c::AbstractVector) = Hyper(Float64.(collect(c)), 0x00)
flags(x::Hyper) = x.flag
dim(x::Hyper) = length(x.c)

# ---- Cayley–Dickson sign table (cached) : e_i · e_j = OM[i,j] · e_{i⊻j} ----
const _OMCACHE = Dict{Int,Matrix{Int}}()
_cdconj(x) = length(x) == 1 ? copy(x) : vcat(_cdconj(@view x[1:end÷2]), -x[end÷2+1:end])
function _cdprod(x, y)
    n = length(x); n == 1 && return x .* y
    h = n ÷ 2
    a, b, c, d = x[1:h], x[h+1:end], y[1:h], y[h+1:end]
    vcat(_cdprod(a, c) .- _cdprod(_cdconj(d), b), _cdprod(d, a) .+ _cdprod(b, _cdconj(c)))
end
function cd_omega(M::Int)
    haskey(_OMCACHE, M) && return _OMCACHE[M]
    OM = zeros(Int, M, M)
    E = [Float64.(1:M .== i) for i in 1:M]
    for i in 1:M, j in 1:M
        v = _cdprod(E[i], E[j]); k = argmax(abs.(v))
        @assert k == ((i-1) ⊻ (j-1)) + 1 "XOR routing broken M=$M ($i,$j)"
        OM[i, j] = Int(sign(v[k]))
    end
    _OMCACHE[M] = OM
end

# ---- structure-tensor multiply and the regular representation L_x ----
function Base.:*(a::Hyper, b::Hyper)
    M = dim(a); OM = cd_omega(M); r = zeros(M)
    for i in 0:M-1, j in 0:M-1
        r[(i ⊻ j) + 1] += OM[i+1, j+1] * a.c[i+1] * b.c[j+1]
    end
    _tot(Hyper(r, a.flag | b.flag))
end
Base.:+(a::Hyper, b::Hyper) = _tot(Hyper(a.c .+ b.c, a.flag | b.flag))
Base.:-(a::Hyper, b::Hyper) = _tot(Hyper(a.c .- b.c, a.flag | b.flag))
Base.:*(s::Real, a::Hyper) = Hyper(s .* a.c, a.flag)

function Lmatrix(x::Hyper)
    M = dim(x); OM = cd_omega(M); L = zeros(M, M)
    for i in 0:M-1, j in 0:M-1
        L[(i ⊻ j) + 1, j + 1] += OM[i+1, j+1] * x.c[i+1]
    end
    L
end
e0(M::Int) = (v = zeros(M); v[1] = 1.0; v)

# ---- totalize output components: never NaN/Inf; saturate → OVER ----
function _tot(x::Hyper)
    c = copy(x.c); f = x.flag
    for i in eachindex(c)
        v = c[i]
        if isnan(v); c[i] = 0.0; f |= SING
        elseif !isfinite(v) || abs(v) > MAXF; c[i] = sign(v) * MAXF; f |= OVER
        end
    end
    Hyper(c, f)
end

# ---- the analytic functions, uniformly = f(L_x) · e0 ----
# forward: exp is entire ⇒ total for every input (zero divisors included)
function hexp(x::Hyper)
    M = dim(x)
    _tot(Hyper(real.(exp(Lmatrix(x)) * e0(M)), x.flag))
end

# inverse-type: guard the single failure point — L_x singular ⇒ flag, don't NaN
function _singular(L)
    s = svdvals(L)
    s[end] <= 1e-9 * max(s[1], 1.0)          # smallest singular value ~ 0
end
# `needs_inverse`: only ops that literally invert (log, x^negative) require L_x nonsingular.
# √ and x^(positive) do NOT — √0 = 0 is fine even though L_0 is singular. So we flag SING
# *only when the matrix function actually fails*, not preemptively (external AI review found
# hsqrt(0) was a false positive: a well-defined op reported as a zero-divisor failure).
function _matfun(f, x::Hyper; needs_inverse::Bool)
    M = dim(x); L = Lmatrix(x); flag = x.flag
    all(iszero, x.c) && return Hyper(zeros(M), flag)   # f(0): resolved directly (√0=0, 0^p=0)
    if needs_inverse && _singular(L)
        return Hyper(zeros(M), flag | SING)            # inversion of a zero divisor: no answer
    end
    v = try
        Y = f(L); Y * e0(M)
    catch
        return Hyper(zeros(M), flag | SING)            # matrix function genuinely failed
    end
    imres = maximum(abs.(imag.(v)))
    rmag  = max(maximum(abs.(real.(v))), 1.0)
    (imres > 1e-8 * rmag) && (flag |= CPLX)            # left the reals ⇒ bigger field
    _tot(Hyper(real.(v), flag))
end
hlog(x::Hyper)  = _matfun(log,  x; needs_inverse=true)    # log(0)=−∞, log(zero-div) singular
hsqrt(x::Hyper) = _matfun(sqrt, x; needs_inverse=false)   # √ never needs inversion
function Base.:^(x::Hyper, p::Real)
    if isinteger(p) && p >= 0
        _matfun(A -> A^Int(p),   x; needs_inverse=false)  # forward, total
    elseif p > 0
        _matfun(A -> A^float(p), x; needs_inverse=false)  # positive fractional: no inversion
    else
        _matfun(A -> A^float(p), x; needs_inverse=true)   # negative power: needs inverse
    end
end

isreal_ok(x::Hyper) = (x.flag & (SING | CPLX)) == 0
function Base.show(io::IO, x::Hyper)
    tag = x.flag == 0 ? "" : "⟦" * (x.flag&SING>0 ? "零因子" : "") *
          (x.flag&CPLX>0 ? "ℂ" : "") * (x.flag&OVER>0 ? "≥" : "") * "⟧"
    print(io, "Hyper", dim(x), "(", join(round.(x.c, digits=4), ","), ")", tag)
end

# ---- self-test: algebraic identities per dimension + the audit regressions ----
function self_test()
    println("HyperTranscend self-test — identities are CHECKED, not assumed")
    rng = Random.MersenneTwister(7)
    approx(a::Hyper, b::Hyper) = maximum(abs.(a.c .- b.c)) < 1e-6
    for M in (1, 2, 4, 8, 16)
        x = Hyper([1.4; 0.25 .* randn(rng, M-1)])       # near identity: log/√ well-conditioned
        oks = [
            ("√x·√x==x",        approx(hsqrt(x)*hsqrt(x), x)),
            ("exp(log x)==x",   approx(hexp(hlog(x)), x)),
            ("x^2==x·x",        approx(x^2, x*x)),
            ("x^0.5·x^0.5==x",  approx((x^0.5)*(x^0.5), x)),
        ]
        s = join(["$n $(v ? "✓" : "✗")" for (n,v) in oks], "  ")
        println("  M=$M: $s")
        @assert all(v for (_,v) in oks)
    end
    # audit regressions (external AI review 2026-07-20): √0 / 0^p must NOT false-flag
    for f in (hsqrt, x->x^0.5, x->x^2.5)
        r = f(Hyper([0.0]))
        @assert flags(r) == 0x00 && r.c[1] == 0.0 "√0 / 0^p wrongly flagged"
    end
    println("  regression √0=0, 0^0.5=0, 0^2.5=0 — no false zero-divisor flag ✓")
    # a genuine zero divisor: forward exp ok, inverse (log) flagged
    z = Hyper([i∈(4,11) ? 1.0 : 0.0 for i in 1:16])
    @assert flags(hexp(z)) & SING == 0 "exp(zero-divisor) should stay total"
    @assert flags(hlog(z)) & SING != 0 "log(zero-divisor) should be flagged"
    println("  zero divisor z=e3+e10: exp forward-total, log flagged ⟦零因子⟧ ✓")
end

end # module

if abspath(PROGRAM_FILE) == @__FILE__
    HyperTranscend.self_test()
end

File: hyper_transcend.py
#!/usr/bin/env python3
# ⚠️ AI-assisted; verify. / 生成AI使用・要検証
"""hyper_transcend — the Python twin of julia/HyperTranscend.jl.

  Experimental unified computation of exp / log / sqrt / ^ / sin / cos / sinh / cosh and a
  linear-ODE mover for a hypercomplex number of any dimension M = 2^k (real → complex →
  quaternion → octonion → sedenion), via the LEFT regular representation:

      a hypercomplex element x  ≙  its left-multiplication matrix  L_x  (M×M),
      and for a function f,      f(x) = f(L_x) · e0.

  Same three-tier discipline as the Julia twin (identical flag vocabulary):
    · SAFE forward  (exp, sin, cos, sinh, cosh, x^{p≥0} via left_power, left_action)
      — total for every input, zero divisors included; a forward power series never inverts.
    · CANDIDATE     (sqrt, log, x^{fractional}) — computed, then the DEFINING IDENTITY is
      verified by a non-recursive residual; trusted only if it holds, else flagged INEXACT
      (a candidate, never a silent lie).  verify_sqrt / verify_log are exposed.
    · BREAK         (log, x^{negative} of a zero divisor) — no unique inverse; named SING.

  Numbers are total: NaN / Inf are named at construction (Hyper()), so no matrix routine can
  crash on a NaN.  Audited by hyper_transcend.py --audit (720 adversarial cases: 0 NaN/Inf,
  0 exceptions, 0 false flags).

  Not a proof that every hypercomplex analytic function is captured — for non-associative
  M ≥ 8 the branch of a matrix function and left-vs-right functions need case-by-case care;
  which identities hold is CHECKED, not assumed (run --test).
"""
import sys, os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np
from scipy.linalg import expm, logm, fractional_matrix_power, funm
from cuda_total import cd_omega                      # the Cayley–Dickson sign table (shared)

SING   = 0x01     # L_x singular ⇒ no unique inverse (zero divisor)
CPLX   = 0x02     # result left the reals (imag residue) ⇒ go to a bigger field
OVER   = 0x04     # a component saturated to ±MAX (range overflow) / NaN at entry
INEXACT = 0x08    # candidate: the defining algebraic identity was NOT verified
MAXF = float(np.finfo(np.float64).max)


class Hyper:
    """A hypercomplex number: c = M real components (M = 2^k) + a uint8 flag.
       Entry totalization: NaN → 0+OVER|SING, ±Inf/out-of-range → ±MAX+OVER."""
    __slots__ = ("c", "flag")
    def __init__(self, c, flag=None):
        v = np.asarray(c, dtype=np.float64).copy()
        if flag is not None:
            self.c, self.flag = v, np.uint8(flag); return
        f = 0
        nan = ~np.isfinite(v)
        if nan.any():
            over = np.isinf(v)
            v = np.where(np.isnan(v), 0.0, v)
            v = np.where(np.isinf(v), np.sign(v) * MAXF, v)
            f |= OVER | (SING if np.isnan(np.asarray(c, dtype=np.float64)).any() else 0)
        big = np.abs(v) > MAXF
        if big.any():
            v = np.where(big, np.sign(v) * MAXF, v); f |= OVER
        self.c, self.flag = v, np.uint8(f)

    def __len__(self): return len(self.c)
    def __repr__(self):
        tag = ""
        if self.flag:
            tag = "⟦" + ("零因子" if self.flag & SING else "") + ("ℂ" if self.flag & CPLX else "") \
                  + ("≥" if self.flag & OVER else "") + ("~" if self.flag & INEXACT else "") + "⟧"
        return f"Hyper{len(self)}({np.round(self.c, 4).tolist()}){tag}"


def e0(M):
    v = np.zeros(M); v[0] = 1.0; return v
def flags(x): return int(x.flag)
def isreal_ok(x): return (x.flag & (SING | CPLX | INEXACT)) == 0

# ---- structure-tensor multiply and the regular representation L_x ----
def _mul(a, b):
    M = len(a); OM = cd_omega(M); r = np.zeros(M)
    for i in range(M):
        for j in range(M):
            r[i ^ j] += OM[i, j] * a.c[i] * b.c[j]
    return _tot(Hyper(r, int(a.flag | b.flag)))

def Lmatrix(x):
    M = len(x); OM = cd_omega(M); L = np.zeros((M, M))
    for i in range(M):
        for j in range(M):
            L[i ^ j, j] += OM[i, j] * x.c[i]
    return L

def _tot(x):
    c = x.c.copy(); f = int(x.flag)
    nan = np.isnan(c)
    if nan.any(): c = np.where(nan, 0.0, c); f |= SING
    ovf = ~np.isfinite(c) | (np.abs(c) > MAXF)
    if ovf.any(): c = np.where(ovf, np.sign(c) * MAXF, c); f |= OVER
    return Hyper(c, f)

def _singular(L):
    if not np.isfinite(L).all(): return True         # non-finite ⇒ treat as singular (guard LAPACK)
    s = np.linalg.svd(L, compute_uv=False)
    return s[-1] <= 1e-9 * max(s[0], 1.0)

def _matfun(f, x, needs_inverse):
    M = len(x)
    if not np.any(x.c): return _apply0(f, M, int(x.flag))     # f(0): resolve directly
    L = Lmatrix(x); flag = int(x.flag)
    if needs_inverse and _singular(L):
        return Hyper(np.zeros(M), flag | SING)                # inversion of a zero divisor
    try:
        v = f(L) @ e0(M)
    except Exception:
        return Hyper(np.zeros(M), flag | SING)                # matrix function genuinely failed
    imres = float(np.max(np.abs(np.imag(v))))
    rmag = max(float(np.max(np.abs(np.real(v)))), 1.0)
    if imres > 1e-8 * rmag: flag |= CPLX                      # left the reals
    return _tot(Hyper(np.real(v), flag))

def _apply0(f, M, flag):
    # f(0): exp(0)=e0 ; sin/sinh(0)=0 ; cos/cosh(0)=e0 ; sqrt(0)=0 ; 0^p=0
    try:
        v = np.real(f(np.zeros((M, M))) @ e0(M))
        return _tot(Hyper(v, flag))
    except Exception:
        return Hyper(np.zeros(M), flag)

# ---- SAFE forward group (total for every input) ----
def hexp(x):   return _matfun(expm, x, False)
def hsin(x):   return _matfun(lambda A: funm(A, np.sin), x, False)
def hcos(x):   return _matfun(lambda A: funm(A, np.cos), x, False)
def hsinh(x):  return _matfun(lambda A: funm(A, np.sinh), x, False)
def hcosh(x):  return _matfun(lambda A: funm(A, np.cosh), x, False)

def left_power(x, n):
    """left power with EXPLICIT bracketing x·(x·(…·1)); sedenions aren't power-associative
       across elements, so the order is fixed and named."""
    n = int(n)
    if n < 0: return _matfun(lambda A: fractional_matrix_power(A, float(n)), x, True)
    acc = Hyper(e0(len(x)), int(x.flag))
    for _ in range(n): acc = _mul(x, acc)
    return acc

def left_action(a, x0, t):
    """solve ẋ = a·x (left action) as x(t) = exp(t·L_a)·x0 — any initial value, any M."""
    v = np.real(expm(t * Lmatrix(a)) @ x0.c)
    return _tot(Hyper(v, int(a.flag | x0.flag)))

# ---- CANDIDATE group: compute, verify a non-recursive identity, else flag INEXACT ----
def _candidate(matf, resid, x, needs_inverse):
    if not np.any(x.c): return _matfun(matf, x, needs_inverse)   # f(0) exact
    y = _matfun(matf, x, needs_inverse)
    if y.flag & SING: return y
    if resid is None: return Hyper(y.c, int(y.flag | INEXACT))
    return y if resid(y) < 1e-6 else Hyper(y.c, int(y.flag | INEXACT))

def hsqrt(x):
    return _candidate(lambda A: fractional_matrix_power(A, 0.5),
                      lambda y: float(np.max(np.abs(_mul(y, y).c - x.c))), x, False)
def hlog(x):
    return _candidate(logm, lambda y: float(np.max(np.abs(hexp(y).c - x.c))), x, True)
def verify_sqrt(x, y): return float(np.max(np.abs(_mul(y, y).c - x.c)))
def verify_log(x, y):  return float(np.max(np.abs(hexp(y).c - x.c)))

def hpow(x, p):
    """x^p. p ∈ ℤ⁺ → exact left_power; p>0 fractional / p<0 → candidate (verify 1/p if integer)."""
    if float(p).is_integer() and p >= 0: return left_power(x, int(p))
    ip = 1.0 / p
    resid = (lambda y: float(np.max(np.abs(left_power(y, int(round(ip))).c - x.c)))) \
            if (float(ip).is_integer() and ip >= 1) else None
    return _candidate(lambda A: fractional_matrix_power(A, float(p)), resid, x, p < 0)


# ---------------------------------------------------------------- self-test / audit
def _approx(a, b): return np.max(np.abs(a.c - b.c)) < 1e-6

def self_test():
    rng = np.random.default_rng(7)
    print("hyper_transcend self-test — identities are CHECKED, not assumed")
    for M in (1, 2, 4, 8, 16):
        x = Hyper(np.concatenate([[1.4], 0.25 * rng.standard_normal(M - 1)]))
        chks = [("√·√==x", _approx(_mul(hsqrt(x), hsqrt(x)), x)),
                ("exp(log)==x", _approx(hexp(hlog(x)), x)),
                ("x^2==x·x", _approx(hpow(x, 2), _mul(x, x))),
                ("x^.5²==x", _approx(_mul(hpow(x, 0.5), hpow(x, 0.5)), x))]
        print(f"  M={M:>2}: " + "  ".join(f"{n} {'✓' if v else '✗'}" for n, v in chks))
        assert all(v for _, v in chks)
    for f in (hsqrt, lambda z: hpow(z, 0.5), lambda z: hpow(z, 2.5)):
        r = f(Hyper(np.zeros(16)))
        assert r.flag == 0 and r.c[0] == 0.0, "√0 / 0^p wrongly flagged"
    z = Hyper([1.0 if i in (3, 10) else 0.0 for i in range(16)])
    assert not (flags(hexp(z)) & SING) and (flags(hlog(z)) & SING)
    x = Hyper(np.concatenate([[0.3], 0.2 * rng.standard_normal(15)]))
    assert _approx(_add(_mul(hsin(x), hsin(x)), _mul(hcos(x), hcos(x))), Hyper(e0(16)))
    a = Hyper(np.concatenate([[0.0], 0.4 * rng.standard_normal(15)]))
    x0 = Hyper(np.concatenate([[1.0], 0.1 * rng.standard_normal(15)]))
    assert _approx(left_action(a, x0, 0.0), x0)
    dt = 1e-6; num = (left_action(a, x0, dt).c - x0.c) / dt
    assert np.max(np.abs(num - _mul(a, x0).c)) < 1e-3
    print("  √0=0, zero-divisor exp/log, sin²+cos²=1, left_action ẋ=a·x ✓")

def _add(a, b): return _tot(Hyper(a.c + b.c, int(a.flag | b.flag)))

def audit():
    """720 adversarial cases: never NaN/Inf, never throw, never false-flag a defined result."""
    rng = np.random.default_rng(0)
    bad_nan = bad_exc = false_flag = total = 0
    huge = MAXF * 2
    cases = [np.zeros(16), [1.0 if i in (3, 10) else 0.0 for i in range(16)],
             1e200 * rng.standard_normal(16), 1e-200 * rng.standard_normal(16),
             np.concatenate([[np.nan], np.zeros(15)]), np.concatenate([[np.inf], np.zeros(15)]),
             rng.standard_normal(16), [-1.0 if i == 0 else 0.0 for i in range(16)],
             [2.0 if i == 0 else 0.0 for i in range(16)]]
    ops = [hexp, hlog, hsqrt, hsin, hcos, hsinh, hcosh,
           lambda z: hpow(z, 2), lambda z: hpow(z, 0.5), lambda z: hpow(z, 2.5),
           lambda z: hpow(z, -1.0), lambda z: left_power(z, 3)]
    for c in cases:
        x = Hyper(c)
        for f in ops:
            total += 1
            try:
                r = f(x)
                if not np.isfinite(r.c).all(): bad_nan += 1; print("  NaN/Inf:", f, x)
            except Exception as e:
                bad_exc += 1; print("  exc:", repr(e)[:60], "on", x)
        total += 1
        try: left_action(Hyper(rng.standard_normal(16)), x, 0.5)
        except Exception as e: bad_exc += 1; print("  exc(left_action):", repr(e)[:60])
    # false-flag: well-defined results must be clean
    for name, cond in [("√0", flags(hsqrt(Hyper(np.zeros(16))))),
                       ("0^2.5", flags(hpow(Hyper(np.zeros(16)), 2.5))),
                       ("exp(zerodiv)", flags(hexp(Hyper([1.0 if i in (3,10) else 0.0 for i in range(16)])))),
                       ("√(4e0)", flags(hsqrt(Hyper([4.0 if i == 0 else 0.0 for i in range(16)]))))]:
        total += 1
        if cond: false_flag += 1; print("  false-flag:", name)
    print(f"総チェック {total} 回 / NaN·Inf {bad_nan} / 例外 {bad_exc} / 誤検出 {false_flag}")
    ok = bad_nan == 0 and bad_exc == 0 and false_flag == 0
    print("★ 全域 かつ 誤検出なし ✓" if ok else "!! 問題あり")
    assert ok


if __name__ == "__main__":
    if "--audit" in sys.argv: audit()
    else: self_test()
